sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/ContainerRepositoryClientBuilder.java

```diff
@@ -0,0 +1,295 @@
+        return this;
+    }
+
+    /**
+     * Adds a custom Http httpPipeline policy.
+     *
+     * @param customPolicy The custom Http httpPipeline policy to add.
+     * @return the ContainerRepositoryClientBuilder.
+     */
+    public ContainerRepositoryClientBuilder addPolicy(HttpPipelinePolicy customPolicy) {
+        Objects.requireNonNull(customPolicy, "'pipelinePolicy' cannot be null.");
+
+        if (customPolicy.getPipelinePosition() == HttpPipelinePosition.PER_CALL) {
+            perCallPolicies.add(customPolicy);
+        } else {
+            perRetryPolicies.add(customPolicy);
+        }
+
+        return this;
+    }
+
+    /**
+     * Builds an instance of ContainerRegistryImpl with the provided parameters.
+     *
+     * @return an instance of ContainerRegistryImpl.
+     */
+    public ContainerRepositoryAsyncClient buildAsyncClient() {
+        if (serializerAdapter == null) {
+            this.serializerAdapter = JacksonAdapter.createDefaultSerializerAdapter();
+        }
+
+        // Service version
+        ContainerRegistryServiceVersion serviceVersion = (version != null)
+            ? version
+            : ContainerRegistryServiceVersion.getLatest();
+
+        if (httpPipeline == null) {
+            this.httpPipeline = createHttpPipeline();
+        }
+
+        ContainerRepositoryAsyncClient client = new ContainerRepositoryAsyncClient(repository, httpPipeline, serializerAdapter, endpoint, serviceVersion.getVersion());
+        return client;
+    }
+
+    private HttpPipeline createHttpPipeline() {
+        Configuration buildConfiguration =
+            (configuration == null) ? Configuration.getGlobalConfiguration() : configuration;
+        if (httpLogOptions == null) {
+            httpLogOptions = new HttpLogOptions();
+        }
+        List<HttpPipelinePolicy> policies = new ArrayList<>();
+
+        policies.add(
+            new UserAgentPolicy(CoreUtils.getApplicationId(clientOptions, httpLogOptions), CLIENT_NAME, CLIENT_VERSION, buildConfiguration));
+        policies.add(new RequestIdPolicy());
+
+        List<HttpHeader> httpHeaderList = new ArrayList<>();
+        if (clientOptions != null) {
+            clientOptions.getHeaders().forEach(header ->
+                httpHeaderList.add(new HttpHeader(header.getName(), header.getValue())));
+        }
+        policies.add(new AddHeadersPolicy(new HttpHeaders(httpHeaderList)));
+
+        policies.addAll(perCallPolicies);
+        HttpPolicyProviders.addBeforeRetryPolicies(policies);
+
+
+        policies.add(retryPolicy == null ? new RetryPolicy() : retryPolicy);
+        policies.add(new CookiePolicy());
+        policies.add(new AddDatePolicy());
+
+        policies.addAll(this.perRetryPolicies);
+        HttpPolicyProviders.addAfterRetryPolicies(policies);
+
+        policies.add(new HttpLoggingPolicy(httpLogOptions));
+
+        if (credential != null) {
+            policies.add(new ContainerRegistryCredentialsPolicy(
```

Comment:
Credential policy should be added after `addBeforeRetryPolicies()` and before `addAfterRetryPolicies()`. No policy should be added after `HttpLoggingPolicy` because we want to log the final state of request before sending it out to the service.

Reply:
The reason I chose to do this is because the credential policy when run also calls into the service endpoint and so I want to make sure all policies that the client is adding are also run when the credential endpoint is called. However, I went ahead and updated to make that possible explicitly.

---
In reply to: [609649064](https://github.com/Azure/azure-sdk-for-java/pull/20368#discussion_r609649064) [](ancestors = 609649064)